### PR TITLE
(#5) test: per-hook numeric-difference study

### PR DESCRIPTION
**Pull request by SieDeta**
## Summary

This PR implements the numeric equivalence study framework, leveraging the shared comparison infrastructure introduced in previous PRs.

**`Merge PR #58 first`**

### Key Changes

* **Shared verdict layer**

  * Reuses `tests.lib.compare` from **PR #58 ** as the single source of truth for numerical comparisons.
  * Implements the **standards-by-mode policy**:

    * **Eager mode:** requires **bitwise equality**
    * **CUDA Graph mode:** uses `allclose(atol=threshold)`
  * Each hook comparison record embeds the shared `Check` structure, including:

    * Maximum absolute difference
    * Mean absolute difference
    * First differing position
    * Detailed comparison metadata

  This ensures that both the numeric study and the test matrix report drift consistently.

* **Integration with PR2's hardened hook isolation**

  * Fully adopts PR2's `isolated_hook` mechanism.
  * Adds a new `--variant compare` mode that executes the `_compare` model inside `isolated_hook(...)`, guaranteeing byte-identical restoration after execution.
  * In this mode, only the selected hook `H` triggers `.copy_()`.
  * The default `--variant p` path continues using the production `_p + hook_selection=H` workflow, which already provides self-isolation.

### `tests/numeric_study.py`

Introduces a CLI utility and `run_study()` implementation that:

* Captures an **unhooked baseline** once.
* Executes **one monitored rollout per hook** using:

  * Clean subprocess execution
  * Full `[N, vocab]` logits capture

For each hook, records:

* Bitwise equality / `allclose` verdict
* Maximum absolute difference
* Mean absolute difference
* First differing token position
* Top-k vocabulary differences at that position
* Greedy token ID divergence

Outputs:

* JSON report
* Human-readable summary table

Failure conditions (non-zero exit status):

* Numerical alert triggered
* Token divergence detected
* Runner error
* Shape mismatch
* Empty capture

**Alert policy:**

* **Eager mode:** any non-bitwise difference
* **CUDA Graph mode:** differences exceeding the configured threshold
* **All modes:** token divergence and runtime failures are always treated as errors

### `tests/test_numeric_study.py`

Adds:

* **12 CPU unit tests**
* **1 GPU smoke test** marked with `numeric`, `gpu`, and `slow`

### Unit Tests

```bash
pytest tests/test_numeric_study.py -m cpu -q
```

Result:

```text
12 passed
```

### Full CPU Test Suite

* Ran the complete CPU suite excluding the **3 pre-existing broken modules**.
* Result: **122 tests passed**, including all tests introduced in this PR.
* CLI `--help` output and test collection both complete cleanly.

### CODE REVIEW

**Comment by Samfisheryu** [comment]
Closing this as superseded by #58. The numeric-study commit from this PR has been cherry-picked into the combined #58 branch for review together with the shared matrix work.
